packages/framework/ai-collab/src/explicit-strategy/agentEditTypes.ts

```diff
@@ -0,0 +1,188 @@
+
+/**
+ * Describes an edit to a field within a node.
+ * @remarks what is the [key: string] for?
+ */
+export interface TreeEditObject {
+	[key: string]: TreeEditValue;
+	[typeField]: string;
+}
+/**
+ * An array of {@link TreeEditValue}'s, allowing a single TreeEdit to contain edits to multiple fields.
+ */
+export type TreeEditArray = TreeEditValue[];
+
+/**
+ * The potential values for a given TreeEdit. These values are typically a field within a node or an entire node,
+ * represented in JSON because they are expected to be returned by an LLM.
+ */
+export type TreeEditValue = JsonPrimitive | TreeEditObject | TreeEditArray;
+
+/**
+ * This is the the final object we expected from an LLM response.
+ * @remarks Because TreeEdit can be multiple different types (polymorphic),
+ * we need to wrap to avoid anyOf at the root level when generating the necessary JSON Schema.
+ */
+export interface EditWrapper {
+	// eslint-disable-next-line @rushstack/no-new-null
+	edit: TreeEdit | null;
+}
+
+/**
+ * Union type representing all possible types of edits that can be made to a tree.
+ */
+export type TreeEdit = SetRoot | Insert | Modify | Remove | Move;
+
+/**
+ * The base interface for all types of {@link TreeEdit}.
+ */
+export interface Edit {
+	explanation: string;
+	type: "setRoot" | "insert" | "modify" | "remove" | "move";
+}
+
+/**
+ * This object provides a way to 'select' either a given node or a range of nodes in an array.
+ */
+export type Selection = ObjectTarget | Range;
+
+/**
+ * A Target object for an {@link TreeEdit}, identified by the target object's Id
+ */
+export interface ObjectTarget {
+	target: string;
+}
+
+/**
+ * Desribes where an object can be inserted into an array.
+ * For example, if an you have an array with 5 objects, and you insert an object at index 3, this differentiates whether you want
+ * the existing item at index 3 should be shifted forward (if the 'place' is 'before') or shifted backwards (if the 'place' is 'after')
+ *
+ * @remarks TODO: Allow support for nested arrays
+ */
+export interface ArrayPlace {
+	type: "arrayPlace";
+	parentId: string;
+	field: string;
+	location: "start" | "end";
+}
+
+/**
+ * Desribes where an object can be inserted into an array.
+ * For example, if an you have an array with 5 objects, and you insert an object at index 3, this differentiates whether you want
+ * the existing item at index 3 should be shifted forward (if the 'place' is 'before') or shifted backwards (if the 'place' is 'after')
+ *
+ * @remarks Why does this and {@link ArrayPlace} exist together?
+ */
+export interface ObjectPlace extends ObjectTarget {
+	type: "objectPlace";
+	// No "start" or "end" because we don't have a way to refer to arrays directly.
+	place: "before" | "after";
+}
+
+/**
+ * A range of objects within an array. This allows the LLM to select multiple nodes at once,
+ * for example during an {@link Remove} operation to remove a range of nodes.
+ */
+export interface Range {
```

Comment:
Am I right to assume this is inclusive of the elements at each end of the range? If so, it might be worth stating that in the comment or perhaps renaming this to something like `InclusiveRange` to be explicit. Some libraries use exclusive ranges, as weird as that might seem.